### DISCUSSION TITLE
Changing dimensions of shape after it's been created.

### DISCUSSION THREAD

**Discussion by SSNTails** [jrouwe/JoltPhysics#1006] [Q&A] [answered]
I understand the errors this potentially introduces depending on the next update of the simulation, but is there a way to adjust the dimensions of a shape after it's been created? For example, modifying the radius of a sphere, or the length/width/height of a box.

As far as use case, I'm specifically thinking of a game where you can change characters. Those characters potentially have different dimensions (one may be taller than the other, or fatter than the other, etc.). Can this be done without destroying the shape and creating a new one?

**Comment by MrOnlineCoder** [comment]
My wild guess is that you can use BodyInterface::SetShape(): https://jrouwe.github.io/JoltPhysics/class_body_interface.html#a051a4064706509e2b1c4dcc6d688676d

So you basically have to create a modified Shape first and then completely overwrite it using SetShape, suspecting that allowing changing body shape dimensions in runtime would affect the optimizations/performance. 

But also note according to https://jrouwe.github.io/JoltPhysics/index.html#creating-shapes:

> Note that after you call Create, the shape is cached and [ShapeSettings](https://jrouwe.github.io/JoltPhysics/class_shape_settings.html) keeps a reference to your shape (see [Memory Management](https://jrouwe.github.io/JoltPhysics/index.html#memory-management)). If you call Create again, the same shape will be returned regardless of what changed to the settings object.

**Comment by jrouwe** [comment, marked as answer]
To change a shape, you have to create a new one and then call BodyInterface::SetShape. Adjusting a shape would not be thread safe (collision queries could be accessing the shape from a different thread) and is therefore not supported. Note that there is ShapeSettings::ClearCachedResult if you want to create a new shape from the same settings object (although with a simple sphere you can also directly construct it).

**Reply by eliasdaler** [reply]
Just to clarify - if I write a single-threaded app (I know that Jolt still does things in multiple threads) - is it safe to change shapes before/after Update is called? Or should I lock/wait for the simulation explicitly somehow?

**Reply by SSNTails** [reply]
Adding/removing is done through the Body Interface, which handles the thread safety for you.

**Reply by jrouwe** [reply]
There is the choice between `PhysicsSystem::GetBodyInterface()` and `PhysicsSystem::GetBodyInterfaceNoLock()`, so you can choose if you want things thread safe or not.

If you're not using multithreading yourself, then you only need to be careful during `PhysicsSystem::Update()`, outside of that call you can use the `NoLock` version.
